### PR TITLE
Explainable ML module

### PR DESCRIPTION
**Pull request by jrzkaminski**
- [x] Explainable module
- [x] Plots

   - Concepts causal effects:
     
image
image

   - CNN Filter Importance:

image
image
image
image

- [x] [Tests](https://github.com/Anaxagor/applybn/tree/explainable-new/tests)
- [x] [Examples](https://github.com/Anaxagor/applybn/tree/explainable-new/examples/explainable)
- [x] [Documentation](https://github.com/Anaxagor/applybn/tree/explainable-new/docs/api/explainable)

TODO:

- [x] Expand Documentation
- [x] Add Progress Bars and more logging via rich

This pull request includes several changes across multiple files to improve the code structure, add new functionality, and enhance the readability of the codebase. The most important changes are grouped into two themes: refactoring and new feature additions.

### Refactoring:
* [`applybn/core/__init__.py`](diffhunk://#diff-4e69f136b35b1252b798a66f447a6c45ecc599e57695fde90cacdb05e5191834L1-R8): Updated import statements to use absolute imports for `exceptions`, `logger`, and `progress_bar` modules.
* [`applybn/explainable/nn_layers_importance/__init__.py`](diffhunk://#diff-fa7bcebea9cbe8333287aa765a15c674690cff81d7ca9f3f8d326d10cca8dcd7L2-R2): Renamed `CNNCausalFilterImportance` to `CausalCNNExplainer` for better clarity.
* [`applybn/feature_extraction/bn_feature_extractor.py`](diffhunk://#diff-bd5c1b4a865c8bafaabac8940a106b13671cca6fbf8eb734a64b6c2f47aa1013R12): Reformatted the `BNFeatureGenerator` class and its methods for better readability and consistency. [[1]](diffhunk://#diff-bd5c1b4a865c8bafaabac8940a106b13671cca6fbf8eb734a64b6c2f47aa1013R12) [[2]](diffhunk://#diff-bd5c1b4a865c8bafaabac8940a106b13671cca6fbf8eb734a64b6c2f47aa1013L27-R32) [[3]](diffhunk://#diff-bd5c1b4a865c8bafaabac8940a106b13671cca6fbf8eb734a64b6c2f47aa1013L42-R51) [[4]](diffhunk://#diff-bd5c1b4a865c8bafaabac8940a106b13671cca6fbf8eb734a64b6c2f47aa1013L75-R91) [[5]](diffhunk://#diff-bd5c1b4a865c8bafaabac8940a106b13671cca6fbf8eb734a64b6c2f47aa1013L105-R119) [[6]](diffhunk://#diff-bd5c1b4a865c8bafaabac8940a106b13671cca6fbf8eb734a64b6c2f47aa1013L124-R138) [[7]](diffhunk://#diff-bd5c1b4a865c8bafaabac8940a106b13671cca6fbf8eb734a64b6c2f47aa1013L143-R161) [[8]](diffhunk://#diff-bd5c1b4a865c8bafaabac8940a106b13671cca6fbf8eb734a64b6c2f47aa1013L184-R204) [[9]](diffhunk://#diff-bd5c1b4a865c8bafaabac8940a106b13671cca6fbf8eb734a64b6c2f47aa1013L195-R216)

### New Feature Additions:
* [`applybn/core/data_iq.py`](diffhunk://#diff-371c95e84c43ce60220a692c135c91929951970cb0b381a53c5fb7fff88c0afeR1-R218): Added a new class `DataIQSKLearn` for computing gold label and true label probabilities, along with several properties for analyzing model confidence and uncertainty.
* [`applybn/explainable/causal_analysis/__init__.py`](diffhunk://#diff-0565500f786b7e66c8a41e925b435786871a4e7edaa97e4b0e01d686bf23161dR1-R6): Added imports for `ConceptCausalExplainer` and `InterventionCausalExplainer` classes.
* [`applybn/explainable/causal_analysis/intervention_causal_effect.py`](diffhunk://#diff-6a3a99eb4684e1529bc6b076e548d0556032dafff2245ef16c636b99df0aed69R1-R234): Added a new class `InterventionCausalExplainer` for estimating feature impact and performing interventions to observe changes in model confidence and uncertainty.

### CODE REVIEW

**Review by Anaxagor — APPROVED** [review]

File: applybn/explainable/causal_analysis/concept_causal_effect.py [hunk 1]
```diff
@@ -0,0 +1,435 @@
...
+
+        Args:
+            D: The dataset for clustering (without index column).
+            num_clusters: The number of clusters to form.
+
+        Returns:
+            Array of cluster labels.
+        """
+        kmeans = KMeans(n_clusters=num_clusters, random_state=42)
+        clusters = kmeans.fit_predict(D)
```
Comment (Roman223) [inline applybn/explainable/causal_analysis/concept_causal_effect.py:65]: Can be a little bit faster with `MiniBatchKMeans`. See: [link](https://scikit-learn.org/stable/modules/clustering.html#mini-batch-kmeans). 

But it probably reduces evaluation time poorly, since it is not a bottleneck....

File: applybn/explainable/causal_analysis/intervention_causal_effect.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+            X_train: Training features.
+            y_train: Training labels.
+            X_test: Test features.
+            y_test: Test labels.
+            n_estimators: Number of estimators for Data-IQ.
+        """
+        self.X_train = X_train.copy()
+        self.y_train = y_train.copy()
+        self.X_test = X_test.copy()
+        self.y_test = y_test.copy()
```
Comment (Roman223) [inline applybn/explainable/causal_analysis/intervention_causal_effect.py:42]: Is it good decision to store data inside object? If it is too big, it can overload system.. Idk whether we consider such cases...

File: applybn/explainable/nn_layers_importance/cnn_filter_importance.py [hunk 3]
```diff
@@ -1,6 +1,471 @@
...
+        """Builds a simple DAG structure, where each layer depends on the previous one."""
+        for idx, (name, layer) in enumerate(self.conv_layers):
+            num_filters = layer.out_channels
+            self.dag[idx] = {
+                "name": name,
+                "layer": layer,
+                "filters": list(range(num_filters)),
+                "parents": [idx - 1] if idx > 0 else [],
+                "children": [idx + 1] if idx < len(self.conv_layers) - 1 else [],
+            }
```
Comment (Roman223) [inline applybn/explainable/nn_layers_importance/cnn_filter_importance.py:81]: What is about dict comprehension here??

File: applybn/explainable/nn_layers_importance/cnn_filter_importance.py [hunk 4]
```diff
@@ -1,6 +1,471 @@
...
+    def _frobenius_norm(tensor: torch.Tensor) -> torch.Tensor:
+        """Default Frobenius norm function that averages over spatial dimensions.
+
+        Args:
+            tensor: A 4D tensor (batch_size, channels, height, width).
+
+        Returns:
+            A 2D tensor (batch_size, channels) with the Frobenius norm computed over spatial dimensions.
+        """
+        return torch.norm(tensor.view(tensor.size(0), tensor.size(1), -1), dim=2)
```
Comment (Roman223) [inline applybn/explainable/nn_layers_importance/cnn_filter_importance.py:309]: Why not build-in func from torch? `torch.norm(..., ord="fro")`

**Review by Roman223 — COMMENTED** [review]
Some snippets should be discussed.

File: applybn/explainable/causal_analysis/concept_causal_effect.py (hunk 1 above)
Comment (jrzkaminski) [reply applybn/explainable/causal_analysis/concept_causal_effect.py:65]: Too insignificant to replace?

File: applybn/explainable/nn_layers_importance/cnn_filter_importance.py (hunk 3 above)
Comment (jrzkaminski) [reply applybn/explainable/nn_layers_importance/cnn_filter_importance.py:81]: Like that?
```suggestion
        self.dag = {
            idx: {
                "name": name,
                "layer": layer,
                "filters": list(range(layer.out_channels)),
                "parents": [idx - 1] if idx > 0 else [],
                "children": [idx + 1] if idx < len(self.conv_layers) - 1 else []
            }
            for idx, (name, layer) in enumerate(self.conv_layers)
        }
```

File: applybn/explainable/nn_layers_importance/cnn_filter_importance.py (hunk 4 above)
Comment (jrzkaminski) [reply applybn/explainable/nn_layers_importance/cnn_filter_importance.py:309]: dim=2 and ord="fro" are equivalent

File: applybn/explainable/causal_analysis/concept_causal_effect.py (hunk 1 above)
Comment (Roman223) [reply applybn/explainable/causal_analysis/concept_causal_effect.py:65]: yep, just mentioned here in case it will be needed in future

File: applybn/explainable/causal_analysis/intervention_causal_effect.py (hunk 2 above)
Comment (jrzkaminski) [reply applybn/explainable/causal_analysis/intervention_causal_effect.py:42]: this is done
